### DISCUSSION TITLE
Compile virgil code to WASM

### DISCUSSION THREAD

**Discussion by kalwalt** [titzer/virgil#235] [Q&A] [answered]
I'm trying to compile some virgil code to WASM, i looked at the mandelbrot example and i would playing around with it. As stated in the Readme https://github.com/titzer/virgil?tab=readme-ov-file#supported-targets WASM is supported.
i run `w3c-wasm Code.v3`
but as a result `v3c-wasm: command not found`
Note that i have `virgil/bin` and `virgil/bin/dev` in PATH, i have also run `aeneas bootstrap`.
Maybe i missed some steps? 
Anyway i can produce a wasm file with the wasi backend with the `v3c-wasi` command but it can not be load in a browser.

**POST EDIT:**
To compile to WASM run for example `v3c -target=wasm -heap-size=500m YourCode.v3` the heap size option is needed to avoid unreachable code.

**Comment by kalwalt** [comment]
Ok i think i get it, you need to run `v3c -target=wasm YourCode.v3` this is correct?

**Reply by kalwalt** [reply]
Note for other virgil users exporting to target `wasm`: It's better to enlarge the heap memory otherwise you can receive unreachable code, use the command:
`v3c -target=wasm -heap-size=500m YourCode.v3`

**Reply by titzer** [reply]
Yeah, that's a good tip. The `wasm` target is also a little less user-friendly in that any safety violations that would normally print a stack traces just trap (typically with unreachable). I.e. there isn't a nice source-level error message.

I should really get around to writing some tutorial pages on how to use the various targets.

**Reply by kalwalt** [reply]
@titzer i saw that you write a new [tutorial for wasm](https://github.com/titzer/virgil/tree/master/doc/tutorial/wasm), many thanks for this! I think it will help many potential users...

**Comment by titzer** [comment, marked as answer]
That's correct. The `wasm` code generator is stable, but none of the platform targets are stable enough to have graduated from `bin/dev/` to `bin/` yet.

**Reply by kalwalt** [reply]
Thank you, this make sense. 🙂

**Comment by gmlewis** [comment]
In addition to the `-heap-size=500m` [suggestion above](https://github.com/titzer/virgil/discussions/235#discussioncomment-9131660) (thank you!)
I've noticed that if I include `lib/util/*.v3` into my build, that I have to add this to my code in order to get it to successfully compile for `-target=wasm`:

```
// Hack to compile to wasm: provide Virgil identifiers that cannot be found:
component System {
  def error(s1: string, s2: string) {}
  def fileClose(file: int) {}
  def fileWriteK(file: int, data: Array<byte>, offset: int, length: int) {}
}
```

**Reply by titzer** [reply]
Thanks, it looks like `lib/util/IO.v3` is responsible for the two file methods and `System.error()` is used in a couple places in `BitMatrix` and `StringBuilder`. TBH the `Buffer` and `Stream` abstractions in the former aren't particularly useful and I've considered removing them; they also predate `Range`, so they could use some improvement.

**Reply by gmlewis** [reply]
While looking for string interpolation, I found the `StringBuilder`, and am using that, but if you have other recommendations, I'm totally open.
Thanks for the cool programming language, @titzer (and your YouTube talks, which is how I found it)!

**Reply by kalwalt** [reply]
> In addition to the `-heap-size=500m` [suggestion above](https://github.com/titzer/virgil/discussions/235#discussioncomment-9131660) (thank you!) I've noticed that if I include `lib/util/*.v3` into my build, that I have to add this to my code in order to get it to successfully compile for `-target=wasm`:
> 
> ```
> // Hack to compile to wasm: provide Virgil identifiers that cannot be found:
> component System {
>   def error(s1: string, s2: string) {}
>   def fileClose(file: int) {}
>   def fileWriteK(file: int, data: Array<byte>, offset: int, length: int) {}
> }
> ```

You are welcome @gmlewis Glenn! Nice taht you find useful this post in some way. note aslo that Ben @titzer has updated the documentation for Wasm compilation, it should be noted on top of this topic. 
And yes this is a cool programming language!

**Reply by gmlewis** [reply]
> note aslo that Ben @titzer has updated the documentation for Wasm compilation, it should be noted on top of this topic.

Yes, however I didn't find your `-heap-size=500m` "trick" in the new documentation either.  😄

**Reply by kalwalt** [reply]
> > note aslo that Ben @titzer has updated the documentation for Wasm compilation, it should be noted on top of this topic.
> 
> Yes, however I didn't find your `-heap-size=500m` "trick" in the new documentation either. 😄

Yes you are right, I don't know if it make sense to update the wasm tutorial but at least i can add a Post Edit note on the very top of this thread. Don't know why we need to set the size of the heap, using emscripten with C/C++ -> Wasm you can add the _sALLOW_MEMORY_GROWTH option, and you are fine. But E,mscripten add alot of code that maybe you will not need.
